### PR TITLE
:zap: add flat fast path to decoder

### PR DESCRIPTION
**Pull request by techouse**
This PR rewrites the decode hot path for flat queries and keeps strict behavior parity for structured/fallback cases.  
It also adds decode snapshot benchmarking and opt-in decode guardrails for Swift + ObjC.

## Why
Recent benchmarks showed this port significantly behind other ports (including Python) on C2/C3 decode workloads.  
Goal: reduce decode latency without changing public API behavior.

### 1. Decode core performance work
- Added a byte-oriented flat-query fast kernel in `Decoder+ParseQuery.swift`.
- Single-pass token scan for common flat cases (no structured syntax).
- Added structured-syntax detection and duplicate-key fallback to full pipeline.
- Preserved key parity semantics:
  - empty-key tokens count toward limits but are excluded from final output
  - parameter/list limit behavior
  - charset sentinel handling
  - `decodeDotInKeys` behavior
  - comma/list overflow semantics
- Added flat finalize shortcut in `Qs+Decode.swift` to avoid unnecessary post-processing for already-concrete flat graphs.

### 2. Behavior/parity test coverage
- Added fast-path parity tests in `DecodeFastPathTests.swift`.
- Kept existing decode behavior tests passing across Swift and ObjC.

### 3. Bench + guardrails
- Added decode snapshot cases C1/C2/C3 to bench output.
- Extended `Bench/scripts/perf_compare.sh` to support decode scenarios.
- Added opt-in decode perf guardrails (Swift + ObjC) and decode baselines.

### 4. Tooling update
- Updated JS fixture dependency used by comparison tooling to `qs@6.15.0`.

## Public API
No public API changes (`Qs.decode`, `QsBridge.decode`, and options surface unchanged).

---

### Main vs PR Head
`main = cbd1d2d`  
`PR head = aec913b`

| Runtime | C1 ms/op | C2 ms/op | C3 ms/op |
|---|---:|---:|---:|
| QsSwift `main` | 4.822 | 440.857 | 444.900 |
| QsSwift `PR head` | 0.037 | 0.472 | 0.536 |
| QsObjC bridge `main` | 4.941 | 438.922 | 443.065 |
| QsObjC bridge `PR head` | 0.075 | 0.818 | 0.886 |

### PR Head vs Other Ports

| Runtime | C1 ms/op | C2 ms/op | C3 ms/op |
|---|---:|---:|---:|
| Node `qs` | 0.064 | 0.851 | 0.866 |
| qs-kotlin | 0.027 | 0.289 | 0.658 |
| QsNet (.NET) | 0.043 | 0.391 | 0.456 |
| `qs.dart` | 0.031 | 0.541 | 0.565 |
| `qs.py` | 0.100 | 1.089 | 1.205 |
| QsSwift `PR head` | 0.037 | 0.472 | 0.536 |
| QsObjC bridge `PR head` | 0.075 | 0.818 | 0.886 |

### Notes
- Node version: `qs@6.14.0`
- Other-port commits:
  - `qs-kotlin`: `f97cbb6`
  - `QsNet`: `122e995`
  - `qs.dart`: `3f8a932`
  - `qs.py`: `cc8e162`

## Summary by CodeRabbit

* **New Features**
  * Added decode snapshot benchmarking (perf-decode, perf-decode-snapshot, perf-all) with committed decode baselines and compare workflow.

* **Bug Fixes**
  * Added explicit DecodeError.emptyDelimiter; tightened decode parity, safety, and list-limit handling; preserve empty segments in comma lists.

* **Performance**
  * Byte-oriented fast path for flat query strings with structured-key scanning and allocation-light tokenization.

* **Tests**
  * New opt-in Swift and ObjC decode performance guardrails and extensive fast-path/edge-case decode tests.

* **Documentation**
  * Updated bench docs, scripts, and baseline references for encode/decode workflows.

* **Chores**
  * Updated JS comparison fixture dependency.

### CODE REVIEW

**Comment by techouse** [comment]
@codex review

File: Sources/QsSwift/Internal/StructuredKeyScan.swift [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+        options: DecodeOptions
+    ) -> StructuredKeyScan {
+        if tempObj.isEmpty { return .empty }
+
+        let allowDots = options.getAllowDots
+        var structuredRoots: Set<String>?
+        var structuredKeys: Set<String>?
+
+        for key in tempObj.keys {
+            let splitAt = firstStructuredSplitIndex(key, allowDots: allowDots)
```
[the remark this answers is not in the document]
Comment (techouse) [reply Sources/QsSwift/Internal/StructuredKeyScan.swift:50]: I would not fix this one. The comment is based on the wrong semantic assumption.

Current behavior is intentional and documented: top-level `%2E` becomes `.` during key decoding, and then splits whenever `allowDots == true`, regardless of `decodeDotInKeys`. `decodeDotInKeys` only changes how encoded dots are treated inside key segments after splitting. That is spelled out in DecodeOptions.swift (line 74) and especially DecodeOptions.swift (line 79).

The existing tests also lock this in:

- DecodeTests.swift (line 3895) expects `a%2Eb=c` to split with `allowDots: true, decodeDotInKeys: false`
- DecodeTests.swift (line 4016) expects the lowercase `%2e` case to do the same

So StructuredKeyScan.swift (line 74) is actually matching the real decode pipeline, and so is the fast-path structured check in Decoder+ParseQuery.swift (line 675). If we gated encoded dots on `getDecodeDotInKeys`, we would misclassify some structured keys as flat and introduce a real regression.

No code change needed here.

**Comment by techouse** [comment]
@codex review

File: Sources/QsSwift/Internal/Decoder+ParseQuery.swift [hunk 2]
```diff
@@ -240,6 +224,757 @@ extension QsSwift.Decoder {
...
+                return options.strictNullHandling ? NSNull() : ""
+            }
+
+            if options.throwOnLimitExceeded, options.listLimit <= 0 {
+                throw DecodeError.listLimitExceeded(limit: options.listLimit)
+            }
+
+            let rhsRange = (tokenRange.lowerBound + splitOffset + 1)..<tokenRange.upperBound
+            if options.comma, rangeContainsByte(bytes, range: rhsRange, byte: 0x2C) {
+                return try decodeCommaListValue(bytes, range: rhsRange, options: options, charset: charset)
```
[the remark this answers is not in the document]
Comment (techouse) [reply Sources/QsSwift/Internal/Decoder+ParseQuery.swift:481]: I checked this against the current implementation and against local Node `qs@6.15.0`, and I’m not applying this change.

With Node `qs`, `throwOnLimitExceeded: true` and `arrayLimit: 0` already throw even for a scalar input:

```js
qs.parse("a=1", { arrayLimit: 0, throwOnLimitExceeded: true })
// throws: "Array limit exceeded. Only 0 elements allowed in an array."
```

So allowing `Qs.decode("a=1", options: .init(listLimit: 0, throwOnLimitExceeded: true))` to succeed would diverge from Node behavior.

The current Swift implementation is intentionally aligned with that behavior in both the generic path and the flat fast path:

- `Decoder+ParseList.swift`
- `Decoder+ParseQuery.swift`

Because strict Node parity is the goal here, I’m keeping the existing behavior and not adding the suggested regression test.

File: Sources/QsSwift/Internal/Decoder+ParseQuery.swift [hunk 3]
```diff
@@ -240,6 +224,757 @@ extension QsSwift.Decoder {
...
+
+        let value: Any = try {
+            if splitOffset == -1 {
+                return options.strictNullHandling ? NSNull() : ""
+            }
+
+            if options.throwOnLimitExceeded, options.listLimit <= 0 {
+                throw DecodeError.listLimitExceeded(limit: options.listLimit)
+            }
+
```
[the remark this answers is not in the document]
Comment (techouse) [reply Sources/QsSwift/Internal/Decoder+ParseQuery.swift:478]: I checked this against the current code and against local Node `qs@6.15.0`, and I’m not applying this change.

With Node `qs`, `throwOnLimitExceeded: true` and `arrayLimit <= 0` already throw even for a plain scalar token, not just when comma-splitting produces a list:

```js
qs.parse("a=1", { arrayLimit: 0, throwOnLimitExceeded: true })
// throws: "Array limit exceeded. Only 0 elements allowed in an array."

qs.parse("a=1", { arrayLimit: -1, throwOnLimitExceeded: true })
// throws: "Array limit exceeded. Only -1 elements allowed in an array."
I checked this against the current code and against local Node `qs@6.15.0`, and I’m not applying this change.

With Node `qs`, `throwOnLimitExceeded: true` and `arrayLimit <= 0` already throw even for a plain scalar token, not just when comma-splitting produces a list:

```js
qs.parse("a=1", { arrayLimit: 0, throwOnLimitExceeded: true })
// throws: "Array limit exceeded. Only 0 elements allowed in an array."

qs.parse("a=1", { arrayLimit: -1, throwOnLimitExceeded: true })
// throws: "Array limit exceeded. Only -1 elements allowed in an array."
```

So narrowing this guard to comma-list decoding only would diverge from Node behavior.

It would also make the flat fast path inconsistent with the existing generic decode path, which already throws broadly for `throwOnLimitExceeded && listLimit <= 0`.

Because strict Node parity is the goal here, I’m keeping the current behavior.

**Comment by techouse** [comment]
@codex review

**Comment by techouse** [comment]
@codex review

**Comment by techouse** [comment]
@codex review

**Comment by techouse** [comment]
@coderabbitai resume

**Comment by techouse** [comment]
@codex review
